README.md
# slackcleaner

Welcome to slack cleaner batch !

If you have a channel full of messages in slack that you want to clear, slack free plan doesn't offer a handy solution.

With their api, we can loop on all the messages to remove them.

Unfortunately, we only can call api once each second to avoid a http rate limit exception. So run the batch, get a coffee and log in slack to see messages disappearing slowly :)

## Prerequisites

For the clean method there is three mandatory arguments for the main method of the launcher : 
(Method name) **token** and **channelId**

# Methods supported
* clean token channelId
* clist token [channelId1,channelId2,...]


# Get your OAuth token
go in your slack menu > manage Apps > build 
Create a slack-app then click on permissions, select following scopes :

- channels:history
- chat:write:user

Save changes then click on Install App to workspace and authorise your own app

Copy / paste your OAuth Access Token. Keep it safe, do not write it on a post-it :)

it should look like this :

```
xoxp-123456789123-123456789123-123456789123-3f2b2d5d123f5d1234b84bf1234efa70
```

# Get your channel id
logged on your slack account, 
browse https://api.slack.com/methods/channels.list/test
click on test method, look at the Json returned and copy the channel's id you want to clean
it should look like this :

```
C3BX5K12B
```

## Contributions
As i'm quite new to open source sharing, please feel free to comment code, do pull requests or just ask for new features



CHANGELOG.md
# Changelog

## v0.3
* Added the clist feature. give all channel ids in parameters and it will count all the messages so you know wich one you should clean !

## v0.2
* Added the support for multiple features on the launcher
* Added JUnit test cases
* Package organization
* Moved config.properties into a resource folder



## v0.1
Well, initialisation and basic clean feature :)

src/test/java/slackCleaner/ContextTest.java
/**
 * 
 */
package slackCleaner;


import org.junit.Assert;
import org.junit.Test;

import com.slackcleaner.launch.Context;



/**
 * @author sleray
 *
 */
public class ContextTest {

	@Test
	public void testContext() {
		Context ctx = new Context();
		//test if the Ctx is not null (obvious)
		Assert.assertNotNull(ctx);
		//test to get at least one propertie from the context
		Assert.assertNotNull(ctx.getApiDelete());
	}

}


src/test/java/slackCleaner/LauncherTest.java
/**
 * 
 */
package slackCleaner;

import static org.junit.Assert.*;

import org.junit.Test;

import com.slackcleaner.launch.Launcher;

/**
 * @author sleray
 *
 */
public class LauncherTest {

	/**
	 * Test method for {@link com.slackcleaner.launch.Launcher#printTitle()}.
	 */
	@Test
	public void testPrintTitle() {
		Launcher.printTitle("JUnit testing");
		//useless JUnit Test but i needed it to test title formatting :)
		assertTrue(true);
	}

}


src/main/java/com/slackcleaner/launch/Clean.java
package com.slackcleaner.launch;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

/**
 * 
 */

/**
 * @author sleray 
 *
 */
public class Clean {
	/**
	 * Loop until there is no more messages returned by slack channels.list method
	 * recursively calls itself
	 * 
	 * @param countMessageDeleted
	 *            to keep track of how many messages are deleted
	 * @throws Exception
	 */

	public void processClean(Integer countMessageDeleted, Context ctx) throws Exception {


		String baseUrl = ctx.getBaseUrl();
		String token = ctx.getToken();
		String apiHistory = ctx.getApiHistory();
		String apiDelete = ctx.getApiDelete();
		String channel = ctx.getChannel();

		String completeListUrl = baseUrl + apiHistory + "?token=" + token + "&channel=" + channel+"&count=1000";
		String jsonResponse = getHTML(completeListUrl);
		JsonObject jsonObject = new JsonParser().parse(jsonResponse).getAsJsonObject();
		JsonArray jsonMessages = jsonObject.get("messages").getAsJsonArray();
		for (String message : listMessages(jsonMessages)) {
			/*
			 * Slack api does not accept more than 1 request per second except a small
			 * burst. To allow us to clean a full channel of thousand message, we'll just
			 * let the batch run and be patient :)
			 */
			Thread.sleep(1001);

			getHTML(baseUrl + apiDelete + "?token=" + token + "&channel=" + channel + "&ts=" + message);
		}
		countMessageDeleted += jsonMessages.size();
		System.out.println("Ok ! (" + countMessageDeleted + " messages deleted yet)");
		boolean hasMore = jsonObject.get("has_more").getAsBoolean();
		if (hasMore) {

			processClean(countMessageDeleted,ctx); 

		} else {
			System.out.println("Job's done !");
		}

	}

	/**
	 * parse the response as a JsonArray to get the timestamp of each message in the
	 * list. this timestamp is the unique key used for delete method in slack
	 * 
	 * @param array
	 *            JSonArray from channels.list
	 * @return list of timestamps as String
	 * @throws Exception
	 */
	private static List<String> listMessages(JsonArray array) throws Exception {
		List<String> messages = new ArrayList<>();
		System.out.print(getTime() + " : Fetching " + array.size() + " messages to delete...");
		for (int i = 0; i < array.size(); i++) {
			JsonObject jsonobject = array.get(i).getAsJsonObject();
			String messageTimeStamp = jsonobject.get("ts").getAsString();
			messages.add(messageTimeStamp);
		}

		return messages;
	}

	/**
	 * @return today's date nicely formatted for logging
	 */
	private static String getTime() {
		SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy hh:mm:ss");
		return sdf.format(new Date());

	}

	/**
	 * This method basically just call a url and get the response as a string
	 * 
	 * @param urlToRead
	 *            build the url from the baseUrl, the api familly and method and
	 *            parameters.
	 * @see slack doc for more info
	 * @return response as String
	 * @throws Exception
	 *             sometimes things go wrong
	 */
	private static String getHTML(String urlToRead) throws Exception {
		StringBuilder result = new StringBuilder();
		URL url = new URL(urlToRead);
		HttpURLConnection conn = (HttpURLConnection) url.openConnection();
		conn.setRequestMethod("GET");
		BufferedReader rd = new BufferedReader(new InputStreamReader(conn.getInputStream()));
		String line;
		while ((line = rd.readLine()) != null) {
			result.append(line);
		}
		rd.close();
		return result.toString();
	}
}


src/main/java/com/slackcleaner/launch/Context.java
/**
 * 
 */
package com.slackcleaner.launch;

import java.util.ResourceBundle;

/**
 * @author sleray
 *
 */
public class Context {

	private  String token;
	private  String channel;
	private  String apiDelete;
	private  String apiInfo ;
	private  String apiHistory ;
	private  String baseUrl ;
	private String version;
	



	/**
	 * Context constructor : pick info from the config.properties file.
	 */
	public Context() {
		ResourceBundle bundle = ResourceBundle.getBundle("config");

		this.baseUrl = bundle.getString("slack.api.baseurl");
		this.apiDelete = bundle.getString("slack.api.chat.delete");
		this.apiHistory = bundle.getString("slack.api.channels.history");
		this.apiInfo = bundle.getString("slack.api.channels.info");
		this.version = bundle.getString("version");
		
	}
	

	/**
	 * @return the apiHistory
	 */
	public String getApiHistory() {
		return apiHistory;
	}


	public void setToken(String token) {
		this.token = token;
	}

	public void setChannel(String channel) {
		this.channel = channel;
	}


	public String getToken() {
		return token;
	}
	
	public String getChannel() {
		return channel;
	}
	
	public String getApiDelete() {
		return apiDelete;
	}
	
	public String getApiInfo() {
		return apiInfo;
	}
	
	public String getBaseUrl() {
		return baseUrl;
	}
	
	public String getVersion() {
		return version;
	}
	

}


src/main/java/com/slackcleaner/launch/Launcher.java
/**
 * 
 */
package com.slackcleaner.launch;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * @author sleray
 *
 */
public class Launcher {

	public static String METHOD_CLEAN = "clean";
	public static String METHOD_CLIST = "clist";

	/**
	 * @param args
	 * @throws Exception
	 */
	public static void main(String[] args) throws Exception {
		// get context from properties file
		Context ctx = new Context();
		ctx.setToken(args[1]);
		printTitle(ctx.getVersion());

		if (METHOD_CLEAN.equals(args[0])) {	
			ctx.setChannel(args[2]);

			Clean clean = new Clean();
			clean.processClean(0, ctx);
		} else if (METHOD_CLIST.equals(args[0])) {
			Clist clist = new Clist();
			//build a list of channelIds
			List<String> listChannel = new ArrayList<>();
			listChannel = new LinkedList<String>(Arrays.asList(args)); 
			listChannel.remove(0); //remove method argument
			listChannel.remove(0); //remove token
			for (String channelId : listChannel) {
				clist.processClist(0, ctx, channelId,null);
			}

		} else {
			System.out.println("Run with those arguments : ");
			System.out.println("Launcher clean token channelId");
			System.out.println("or : ");
			System.out.println("Launcher clist token [channelId1, channelId2,...]");
		}

	}

	public static void printTitle(String version) {
		System.out.println("    _____ _            _     _____ _                ");            
		System.out.println("   / ____| |          | |   / ____| |          ");                 
		System.out.println("  | (___ | | __ _  ___| | _| |    | | ___  __ _ _ __   ___ _ __"); 
		System.out.println("   \\___ \\| |/ _` |/ __| |/ / |    | |/ _ \\/ _` | '_ \\ / _ \\ '__|");
		System.out.println("   ____) | | (_| | (__|   <| |____| |  __/ (_| | | | |  __/ |");   
		System.out.println("  |_____/|_|\\__,_|\\___|_|\\_\\\\_____|_|\\___|\\__,_|_| |_|\\___|_| ... "+version); 
		
	}

}


src/main/java/com/slackcleaner/launch/Clist.java
package com.slackcleaner.launch;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

/**
 * 
 */

/**
 * @author sleray 
 *
 */
public class Clist {
	/**
	 * Loop until there is no more messages returned by slack channels.list method
	 * recursively calls itself
	 * 
	 * @param countMessage
	 *            to keep track of how many messages are counted
	 * @throws Exception
	 */

	public void processClist(Integer countMessage, Context ctx, String channelId, String oldest_ts) throws Exception {


		String baseUrl = ctx.getBaseUrl();
		String token = ctx.getToken();
		String apiHistory = ctx.getApiHistory();
		String apiInfo = ctx.getApiInfo();
		String completeListUrl = "";
		if (oldest_ts == null) {
			//first call we get the 1000 most recents messages.
		 completeListUrl = baseUrl + apiHistory + "?token=" + token + "&channel=" + channelId+"&count=1000";
		} else {
			//next calls we get the next 1000 messages
			completeListUrl = baseUrl + apiHistory + "?token=" + token + "&channel=" + channelId+"&count=1000&latest="+oldest_ts;	
		}
		String jsonResponse = getHTML(completeListUrl);
		JsonObject jsonObject = new JsonParser().parse(jsonResponse).getAsJsonObject();
		JsonArray jsonMessages = jsonObject.get("messages").getAsJsonArray();


		
		countMessage += jsonMessages.size();
		//get the timestamp of the oldest message to iterate through the collection with a pagination of 1000
		
		boolean hasMore = jsonObject.get("has_more").getAsBoolean();
		/*
		 * Slack api does not accept more than 1 request per second except a small
		 * burst. To allow us to clean a full channel of thousand message, we'll just
		 * let the batch run and be patient :)
		 */
		Thread.sleep(1001);
		if (hasMore) {
			String actualOldestTs = jsonMessages.get(jsonMessages.size()-1).getAsJsonObject().get("ts").getAsString();
			//For channels above 10K messages, a little animation to wait.
			System.out.print(".");
			processClist(countMessage,ctx,channelId,actualOldestTs); 

		} else {
			//TODO insert basic infos on the channel (title, purpose...)
			completeListUrl = baseUrl + apiInfo + "?token=" + token + "&channel=" + channelId;
			jsonResponse = getHTML(completeListUrl);
			jsonObject = new JsonParser().parse(jsonResponse).getAsJsonObject();
			
			String channelName = jsonObject.get("channel").getAsJsonObject().get("name").getAsString();
			
			System.out.println("Channel Info for '"+channelName+"' ("+channelId+")");
			System.out.println(countMessage+" messages counted");
		}

	}

	
	/**
	 * This method basically just call a url and get the response as a string
	 * 
	 * @param urlToRead
	 *            build the url from the baseUrl, the api familly and method and
	 *            parameters.
	 * @see slack doc for more info
	 * @return response as String
	 * @throws Exception
	 *             sometimes things go wrong
	 */
	private static String getHTML(String urlToRead) throws Exception {
		StringBuilder result = new StringBuilder();
		URL url = new URL(urlToRead);
		HttpURLConnection conn = (HttpURLConnection) url.openConnection();
		conn.setRequestMethod("GET");
		BufferedReader rd = new BufferedReader(new InputStreamReader(conn.getInputStream()));
		String line;
		while ((line = rd.readLine()) != null) {
			result.append(line);
		}
		rd.close();
		return result.toString();
	}
}
